exercises/practice/state-of-tic-tac-toe/.meta/src/reference/java/StateOfTicTacToe.java

```diff
@@ -0,0 +1,111 @@
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
+
+class StateOfTicTacToe {
+    public GameStates determineState(String[] board) {
+        int xWin = 0;
+        int oWin = 0;
+
+        List<String> columns = getColumns(board);
+        List<String> diagonals = getDiagonals(board);
+        
+        int xCount = count('X', board);
+        int oCount = count('O', board);
+
+        if (oCount > xCount) {
+            throw new IllegalArgumentException("Wrong turn order: O started");
+        }
+
+        if (xCount > oCount + 1) {
+            throw new IllegalArgumentException("Wrong turn order: X went twice");
+        }
+
+        List<String> allCombos = new ArrayList<>();
+        
+        allCombos.addAll(Arrays.asList(board));
+        allCombos.addAll(columns);
+        allCombos.addAll(diagonals);
+        
+        for (String row : allCombos) {
```

Comment:
I'm wondering whether it makes sense to simplify this `for` loop, since I believe the following would be sufficient:

```java
var winCount = allCombos
    .stream()
    .filter(x -> x.equals("XXX") || x.equals("OOO"))
    .count();

if (winCount > 1) {
    throw new IllegalArgumentException("Impossible board: game should have ended after the game was won");
}

if (winCount == 1) {
    return GameStates.WIN;
}

if (xCount + oCount < 9) {
    return GameStates.ONGOING;
}

return GameStates.DRAW;

Reply:
Disregard this suggestion due to the thread above.